app/src/main/java/bg/zahov/app/ui/history/HistoryViewModel.kt

```diff
@@ -1,44 +1,97 @@
-    }
-    private val serviceError by lazy {
-        application.getServiceErrorProvider()
-    }
-    private val _state = MutableLiveData<State>(State.Default)
-    val state: LiveData<State>
-        get() = _state
+/**
+ * @property workouts A list of past workouts to be displayed.
+ */
+data class HistoryUiState(
+    val workouts: List<HistoryWorkout> = listOf()
+)
+
+/**
+ * @property duration The duration of the workout in a string format (e.g., "00:30:00").
+ * @property volume The total volume of weight lifted during the workout.
+ * @property date The date when the workout was performed.
+ * @property exercises exercises performed in the workout( 3 x exerciseName - example).
+ * @property bestSets best sets for the workout(the ones where the most weight was lifted).
+ * @property personalRecords number of personal records set during the workout.
+ */
+data class HistoryWorkout(
+    val id: String,
+    val name: String,
+    val duration: String,
+    val volume: String,
+    val date: String,
+    val exercises: List<String>,
+    val bestSets: List<String>,
+    val personalRecords: String
+)
+
+/**
+ * This ViewModel handles the fetching of past workouts and updates the UI state accordingly.
+ *
+ * @property workoutProvider An instance of [WorkoutProvider] used to retrieve past workouts.
+ * @property serviceError An instance of [ServiceErrorHandler] used for handling errors during data retrieval.
+ */
+class HistoryViewModel(
+    private val workoutProvider: WorkoutProvider = Inject.workoutProvider,
+    private val serviceError: ServiceErrorHandler = Inject.serviceErrorHandler
+) : ViewModel() {
+    private val _uiState = MutableStateFlow(HistoryUiState())
+    val uiState: StateFlow<HistoryUiState> = _uiState
 
     init {
         viewModelScope.launch {
-            _state.postValue(State.Loading(View.VISIBLE, View.GONE))
             workoutProvider.getPastWorkouts().collect {
                 try {
-                    _state.postValue(State.Data(it.sortedByDescending { item -> item.date }))
+                    _uiState.update { old ->
+                        old.copy(it.map { workout -> workout.toHistoryWorkout() })
+                    }
                 } catch (e: CriticalDataNullException) {
                     serviceError.initiateCountdown()
                 }
             }
         }
     }
+}
 
-    sealed interface State {
-        object Default : State
-        data class Loading(val loadingVisibility: Int, val workoutsVisibility: Int) : State
-        data class Data(val workouts: List<Workout>) : State
-    }
-}
\ No newline at end of file
+/**
+ * Converts a [Workout] object to a [HistoryWorkout] object.
+ *
+ * This extension function transforms a [Workout] instance into a [HistoryWorkout] instance by extracting
+ * relevant data and formatting it appropriately.
+ *
+ * @return A [HistoryWorkout] instance populated with data from this [Workout].
+ */
+fun Workout.toHistoryWorkout(): HistoryWorkout {
```

Comment:
This parser operation can be applied in the provider on all workouts (in the background thread) before serving them to the ViewModel.